### ISSUE TITLE
Bug in Gdal.Polygonize results in missing polygon area?

### ISSUE THREAD

**Issue by WindingWinter** [OSGeo/gdal#1158]
I'm testing this on GDAL 2.3.1, Windows 10 64 bit

I believe that there is a bug in `Gdal.Polygonize`. I'm using Gdal.Net but I think one can easily reproduce the same problem in python binding or native C++.

There are two raster files inside the zip file: 

1. `loganw50.tiff`
2. `loganw150.tiff`

Run the two files with the following `polygonize` code:

```
        private static void Polygonize(string rasterFilePath)
        {
             var ds = Gdal.Open(rasterFilePath, Access.GA_ReadOnly);
            Debug.Assert(ds.RasterCount == 1);
            var band = ds.GetRasterBand(1);
            var vectorFilePath = Path.ChangeExtension(rasterFilePath, "shp");
            var ogrDriver = Ogr.GetDriverByName("ESRI Shapefile");
            var dst_ds = ogrDriver.CreateDataSource(vectorFilePath, null);
            var dst_layer = dst_ds.CreateLayer("Polygon", null, wkbGeometryType.wkbCompoundCurve, null); 

            var fieldDef = new FieldDefn("CellInfo", FieldType.OFTInteger);
            dst_layer.CreateField(fieldDef, 0);

            var outcome = Gdal.Polygonize(band, null, dst_layer, 0, null, null, null);
            var polygonArea = GetPolygonArea(dst_layer);
            var rasterArea = GetRasterArea(ds);
           
            var percentageDiff = Math.Abs((polygonArea - rasterArea) * 100 / rasterArea); //compute for missing area, when the algorithm is working perfectly it should have only at most rounding off errors as percentage difference. 
        }

        private static double GetPolygonArea(Layer dst_layer)
        {
            dst_layer.ResetReading();

            Feature feature;
            double polygonArea = 0;
            while ((feature = dst_layer.GetNextFeature()) != null)
            {
                var cellVal = feature.GetFieldAsInteger("CellInfo");
                var geometry = feature.GetGeometryRef();
                polygonArea += Math.Abs(geometry.GetArea());

            }

            return polygonArea;
        }

        private static double GetRasterArea(Dataset ds)
        {
            var nCol = ds.RasterXSize;
            var nRow = ds.RasterYSize;
            double[] geoTransform = new double[6];
            ds.GetGeoTransform(geoTransform);
            var cellSize = Math.Abs(geoTransform[5]);
            Band band = ds.GetRasterBand(1);
            var allValues = new int[nCol * nRow];
            double noVal;
            int hasVal;
            band.GetNoDataValue(out noVal, out hasVal);
            band.ReadRaster(0, 0, nCol, nRow, allValues, nCol, nRow, 0, 0);
            var goValueCount = allValues.Count(val => val != noVal);
            return goValueCount * cellSize * cellSize;

        }

```

There are missing areas for both cases, with  `loganw50.tiff` has around 0.55% missing areas, whereas  `loganw150.tiff` has around 0.08% missing areas.

I don't think the difference between raster and vectorize polygons can be explained by rounding off errors ( if there is any), it must be a bug. Because the two files have different missing areas percentage, and because when I render it, it shows the holes in the output.

For both shapefile outputs, there are "holes" in the watershed-- `loganw50` has more holes than `loganw150`, as shown below ( ignore the streamlines, they aren't important):

![threshold50](https://user-images.githubusercontent.com/13161574/50198244-ada02c80-0385-11e9-8151-ef3dc7ce8078.png)
This is for `loganw50.tiff`

![threshold150](https://user-images.githubusercontent.com/13161574/50198258-bb55b200-0385-11e9-8481-d3f257c0910f.png)
This is for `loganw150.tiff`

I have another file with close to 0% difference, not shown here. 
[Loganw.zip](https://github.com/OSGeo/gdal/files/2693741/Loganw.zip)

**Comment by rouault** [comment]
Looking a bit, I don't get holes when displaying with QGIS, but at the locations where you see holes, the polygons have self intersections and don't pass the ST_IsValid() test. When looking closer at the hole towards the top of the image, ![zoom](https://user-images.githubusercontent.com/1192433/50401969-f34c6300-0792-11e9-87df-b3867d52e2bb.png), I can see that the polygon contains a single inner ring for the 3 green small interior squares, and thus this ring is self-intersecting. The polygonizer should rather emit 3 distinct inner rings

```
 ogrinfo out.gpkg -sql "select geom, st_isvalidreason(geom) from out where fid = 86" -geom=summary -al -q

Layer name: SELECT
OGRFeature(SELECT):0
  st_isvalidreason(geom) (String) = Ring Self-intersection[448484.01909127 4659392.44691631]
  POLYGON : 188 points, 1 inner rings (13 points)
```

**Comment by rouault** [comment]
Simple reproducer. Given test.asc:
```
ncols        4
nrows        4
xllcorner    0
yllcorner    0
cellsize     1
 1 1 1 1
 1 0 1 1
 1 1 0 1
 1 1 1 1
```
```
$ python swig/python/scripts/gdal_polygonize.py test.asc test.gpkg
$ ogrinfo -sql "select geom, st_isvalidreason(geom) from out" test.gpkg -q

Layer name: SELECT
OGRFeature(SELECT):0
  st_isvalidreason(geom) (String) = Ring Self-intersection[2 2]
  POLYGON ((0 4,0 0,4 0,4 4,0 4),(1 3,2 3,2 2,3 2,3 1,2 1,2 2,1 2,1 3))

OGRFeature(SELECT):1
  st_isvalidreason(geom) (String) = Valid Geometry
  POLYGON ((1 3,1 2,2 2,2 3,1 3))

OGRFeature(SELECT):2
  st_isvalidreason(geom) (String) = Valid Geometry
  POLYGON ((2 2,2 1,3 1,3 2,2 2))
```

**Comment by WindingWinter** [comment]
> Simple reproducer. Given test.asc:
> 
> ```
> ncols        4
> nrows        4
> xllcorner    0
> yllcorner    0
> cellsize     1
>  1 1 1 1
>  1 0 1 1
>  1 1 0 1
>  1 1 1 1
> ```
> ```
> $ python swig/python/scripts/gdal_polygonize.py test.asc test.gpkg
> $ ogrinfo -sql "select geom, st_isvalidreason(geom) from out" test.gpkg -q
> 
> Layer name: SELECT
> OGRFeature(SELECT):0
>   st_isvalidreason(geom) (String) = Ring Self-intersection[2 2]
>   POLYGON ((0 4,0 0,4 0,4 4,0 4),(1 3,2 3,2 2,3 2,3 1,2 1,2 2,1 2,1 3))
> 
> OGRFeature(SELECT):1
>   st_isvalidreason(geom) (String) = Valid Geometry
>   POLYGON ((1 3,1 2,2 2,2 3,1 3))
> 
> OGRFeature(SELECT):2
>   st_isvalidreason(geom) (String) = Valid Geometry
>   POLYGON ((2 2,2 1,3 1,3 2,2 2))
> ```

I wonder whether is it possible to implement the checking of polygon validity inside the polygonize function? The function can then warn the user if the output is not correct and the user can then submit the report for your investigation.

**Comment by WindingWinter** [comment]
That would be good; any timeline for the fix to be available?

**Comment by WindingWinter** [comment]
Just wonder whether this problem is already fixed  in [GDAL 3.0.0](/OSGeo/gdal/releases/tag/v3.0.0)?

**Comment by mloskot** [comment]
Why don't you try out? If it's not fixed, then consider debugging the issue and submitting pulling request yourself or go back to square https://github.com/OSGeo/gdal/issues/1158#issuecomment-450100880
